homeassistant/components/modbus/sensor.py

```diff
@@ -110,7 +110,9 @@ async def async_update(self, now: datetime | None = None) -> None:
         result = self.unpack_structure_result(raw_result.registers)
         if self._coordinator:
             if result:
-                result_array = list(map(int, result.split(",")))
+                result_array = list(
+                    map(float if self._precision else int, result.split(","))
```

Comment:
Each value in result is a 16bit register, so this change converts 16bit to a float, but a float needs minimum 32bits.

Reply:
Not true ... you have missed the handling in the base_platform.py:
v_temp = self._scale * entry + self._offset

If SCALE is a float then the result will be a float, even if the register is an integer

Reply:
But you are mixing issues..."entry" is derived from val, which originates from struct.unpack (line 174), and thus do not originate from the raw registers.

Reply:
I have a result where anything below 0.025kw should be treated as 0. How do you suggest this is achieved?

Reply:
No I did not miss that, but using the "float" have a number of bugs, which is why we made slaves instead. The list is just waiting to be removed, so I am reluctant to accept changes, unless you have a very good use case (that cannot be solved with slaves).

Reply:
Correct .. it results from a comma delineated string that will have floats in if the base class has a non integer scale.
It's the result of the function with min max checking in we are discussing on other pr

Reply:
Yes but you should use slaves instead.

Reply:
I am using slaves and getting an exception that this fixes

Reply:
Which exception ? (you have not included any log that show that). In this case it is important to add a test case, that so we can see that it works with your PR (and do not work without).

Reply:
Will do. Set scale to 0.025 and precision to 2. Bang!

Reply:
If the code causes an exception, then it is a different matter, because then your PR is a bug fix and not an enhancement, but it is nice to have the exception documented for future reference.

Reply:
Attach a log please.

Reply:
config:
        precision: 3
        scale: 0.005


sensor.py test code:
    async def async_update(self, now: datetime | None = None) -> None:
        ...

        result = self.unpack_structure_result(raw_result.registers)
        if self._coordinator:
            if result:
                try:
115:                result_array = list(map(int, result.split(",")))
                except Exception:
                    _LOGGER.exception(f"{result=}")
                    raise

Exception:
 File "/config/custom_components/modbus2/sensor.py", line 115, in async_update
    result_array = list(map(int, result.split(",")))
ValueError: invalid literal for int() with base 10: '0.040'